src/mbgl/style/conversion/geojson_options.cpp

```diff
@@ -83,12 +87,62 @@ optional<GeoJSONOptions> Converter<GeoJSONOptions>::operator()(const Convertible
         if (toBool(*lineMetricsValue)) {
             options.lineMetrics = *toBool(*lineMetricsValue);
         } else {
-            error = { "GeoJSON source lineMetrics value must be a boolean" };
+            error.message = "GeoJSON source lineMetrics value must be a boolean";
+            return nullopt;
+        }
+    }
+
+    const auto clusterProperties = objectMember(value, "clusterProperties");
+    if (clusterProperties) {
+        if (!isObject(*clusterProperties)) {
+            error.message = "GeoJSON source clusterProperties value must be an object";
+            return nullopt;
+        }
+        GeoJSONOptions::ClusterProperties result;
+        eachMember(
+            *clusterProperties,
+            [&](const std::string& k,
+                const mbgl::style::conversion::Convertible& v) -> optional<conversion::Error> {
+                // Each property shall be formed as ["key" : [operator, [mapExpression]]]
+                // or ["key" : [[operator, ['accumulated'], ['get', key]], [mapExpression]]]
+                if (!isArray(v) || arrayLength(v) != 2) {
+                    error.message =
+                            "GeoJSON source clusterProperties member must be an array with length of 2";
+                    return {};
+                }
+                auto map = expression::dsl::createExpression(arrayMember(v, 1));
+                std::unique_ptr<expression::Expression> reduce;
+                if(isArray(arrayMember(v, 0))){
+                    reduce = expression::dsl::createExpression(arrayMember(v, 0));
+                }
+                else{
+                    auto reduceOp = toString(arrayMember(v, 0));
+                    if (!reduceOp) {
+                        error.message =
+                                "GeoJSON source clusterProperties member must contain a valid operator";
+                        return {};
+                    }
+                    std::stringstream ss;
+                    // Reformulate reduce expression to [operator, ['accumulated'], ['get', key]]
+                    ss << std::string(R"([")") << *reduceOp
```

Comment:
Can we build expression without parsing string? Something like:

`reduce = createCompoundExpression(*reduceOp, {expression::dsl::accumulated(), expression::dsl::get(k)});`